pkg/ipam/crd.go

```diff
@@ -418,29 +417,31 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
 		}
 		for _, a := range n.allocators {
 			if a.family == ipFamily {
-				allocator = a
+				markedToReleaseIPMap[ip] = a
 			}
 		}
-		if allocator == nil {
+		if _, ok := markedToReleaseIPMap[ip]; !ok {
 			continue
 		}
+		releaseUpstreamSyncNeeded = true
+	}
+	n.mutex.Unlock()
 
+	for ip, allocator := range markedToReleaseIPMap {
 		// Some functions like crdAllocator.Allocate() acquire lock on allocator first and then on nodeStore.
 		// So release nodestore lock before acquiring allocator lock to avoid potential deadlocks from inconsistent
 		// lock ordering.
-		n.mutex.Unlock()
-		allocator.mutex.Lock()
-		_, ok := allocator.allocated[ip]
-		allocator.mutex.Unlock()
+		allocator.mutex.RLock()
 		n.mutex.Lock()
 
-		if ok {
+		if _, ok := allocator.allocated[ip]; ok {
 			// IP still in use, update the operator to stop releasing the IP.
 			n.ownNode.Status.IPAM.ReleaseIPs[ip] = ipamOption.IPAMDoNotRelease
 		} else {
 			n.ownNode.Status.IPAM.ReleaseIPs[ip] = ipamOption.IPAMReadyForRelease
 		}
-		releaseUpstreamSyncNeeded = true
+		n.mutex.Unlock()
```

Comment:
Add a comment to explain why we are not deferring it anymore: 
```suggestion
                 // unlock node mutex before the allocator mutex to avoid IP addresses being allocated to newly created pods during the unlocked period
		n.mutex.Unlock()
```